### PR TITLE
Kinesis stream of upload actions

### PR DESCRIPTION
**Pull request by mbarton**
This PR adds a kinesis stream of actions to be performed as parts of a video are uploaded.

The original design (#320) used a lambda triggered by S3 PutObject events. Unfortunately, the ordering of the events received was not always correct, leading to problems when hooked up to YouTube which does not like receiving part 2 before part 2 etc.

This PR removes the S3 lambda and sends the upload complete message on Kinesis once the client has 'completed' the part, in a separate POST request. Once all parts are complete, we trigger an S3 multipart copy to create the `full` part which can be ingested by Pluto.

`UploadAction` subclasses express the operations sent on the stream:

- `UploadPartToYouTube`: currently just logs, subsequent PR will add the functionality
- `DeleteParts`: removes parts, leaving just the `full` part

Potentially unhelpful architecture diagram:

manual complete uploading

### CODE REVIEW

File: common/src/main/scala/com/gu/media/upload/UploadsTable.scala [hunk 1]
```diff
@@ -4,17 +4,10 @@ import com.gu.media.aws.DynamoAccess
 import com.gu.scanamo.syntax._
```
Comment (Reettaphant) [inline common/src/main/scala/com/gu/media/upload/UploadsTable.scala]: Is there some reason why this is not called a datastore like the atom and audit datastores?

File: app/controllers/UploadController.scala [hunk 2]
```diff
@@ -46,41 +51,54 @@ class UploadController(val authActions: HMACAuthActions, awsConfig: AWSConfig, o
...
+      NoContent
     }
   }
 
-  private def buildUpload(atom: MediaAtom, user: User, size: Long) = {
-    val id = UUID.randomUUID().toString
-
-    Upload(
-      id = id,
+  private def buildUpload(atom: MediaAtom, channelId: String, user: User, size: Long) = {
```
Comment (akash1810) [inline app/controllers/UploadController.scala]: could we get the `channelId` from the `atom`?

File: cloudformation/media-atom-maker-dev.yml [hunk 3]
```diff
@@ -240,17 +240,26 @@ Resources:
                     - "arn:aws:dynamodb:${Region}:${Account}:table/${Table}"
                     - { Region: !Ref "AWS::Region", Account: !Ref "AWS::AccountId", Table: !Ref "UploadTrackingDynamoTable" }
               - Effect: "Allow"
-                Action: ["s3:GetObject"]
+                Action: ["s3:*"]
```
Comment (akash1810) [inline cloudformation/media-atom-maker-dev.yml]: why this change?

File: common/src/main/scala/com/gu/media/aws/UploadAccess.scala [hunk 4]
```diff
@@ -12,6 +12,10 @@ trait UploadAccess { this: Settings with AwsAccess =>
   lazy val uploadSTSClient = createUploadSTSClient()
 
   private def createUploadSTSClient() = {
+    if(!userUploadRole.startsWith("arn:")) {
```
Comment (akash1810) [inline common/src/main/scala/com/gu/media/aws/UploadAccess.scala]: why do we need to do this?

Comment (mbarton) [inline common/src/main/scala/com/gu/media/aws/UploadAccess.scala]: If you use the non-arn format you only get an "Unauthorized" error from Amazon which is quite confusing. After the second time it happened to me, I added a check to give a better exception 😄

File: app/controllers/UploadController.scala (hunk 2 above)
Comment (mbarton) [inline app/controllers/UploadController.scala]: we do, I just pattern match on it at the level above since we need to issue an error if it is missing

File: cloudformation/media-atom-maker-dev.yml (hunk 3 above)
Comment (mbarton) [inline cloudformation/media-atom-maker-dev.yml]: To allow the deletion of entries once uploads are complete. I'll make the permissions more specific to be clearer

File: app/controllers/UploadController.scala (hunk 2 above)
Comment (mbarton) [inline app/controllers/UploadController.scala]: I'll change it so we pattern match inside the method and throw an exception (caught in the `parse` wrapper method) if it isn't there

Comment (mbarton) [inline app/controllers/UploadController.scala]: Accidentally fixed in #326 instead of this PR

File: common/src/main/scala/com/gu/media/upload/UploadLifecycle.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  private def chunk(uploadId: String, size: Long): List[UploadPart] = {
+    val boundaries = Upload.calculateChunks(size)
+
+    boundaries.zipWithIndex.map { case ((start, end), id) =>
+      UploadPart(UploadPartKey(aws.userUploadFolder, uploadId, id).toString, start, end)
+    }
+  }
+
+  private def createFullObject(upload: Upload): UploadFullKey = {
```
Comment (jennysivapalan) [inline common/src/main/scala/com/gu/media/upload/UploadLifecycle.scala]: I'm not sure what `createFullObject` could it be renamed to something more descriptive?

Comment (mbarton) [inline common/src/main/scala/com/gu/media/upload/UploadLifecycle.scala]: I renamed it to `createCompleteObject` and the corresponding URI class to `CompleteUploadKey`
